File: backend/app/api/Routes/media.py
from flask import g, jsonify, request, Blueprint
from sqlalchemy import func, select, and_, or_
from database.models import Media, MediaCategory, Supplier, Product, ProductCategory, Quantity, ChildProduct
from marshmallow import ValidationError
from app.api.Schemas.media_schema import MediaSchema, MediaCategorySchema

media_bp = Blueprint("media", __name__)
media_schema = MediaSchema()
media_category_schema = MediaCategorySchema(many=True)

ProductCategory_id = 4


# --- GET all Media ---
@media_bp.route("/media", methods=["GET"])
def get_media():
    db = g.db
    results = db.execute(select(Media)).scalars().all()
    return jsonify([item.to_dict(include_relationships=True) for item in results]), 200


# --- GET media categories (id + name) ---
@media_bp.route("/media_categories", methods=["GET"])
def get_media_categories():
    db = g.db
    categories = db.execute(select(MediaCategory)).scalars().all()
    return jsonify(media_category_schema.dump(categories)), 200


# --- GET single Media item ---
@media_bp.route("/media/<int:id>", methods=["GET"])
def get_media_item(id):
    db = g.db
    item = db.get(Media, id)
    if not item:
        return jsonify({"error": "Media item not found"}), 404
    return jsonify(item.to_dict(include_relationships=True)), 200


# --- POST new Media item ---
@media_bp.route("/media", methods=["POST"])
def create_media():
    db = g.db
    try:
        data = media_schema.load(request.get_json())
    except ValidationError as err:
        return jsonify({"errors": err.messages}), 400

    supplier = db.get(Supplier, data["supplier_id"])
    if not supplier:
        return jsonify({"error": "Invalid supplier ID"}), 400

    category = db.get(MediaCategory, data["category_id"])
    if not category:
        return jsonify({"error": "Invalid category ID"}), 400

    # 1️⃣ Create Media record
    new_media = Media.from_dict(data)
    db.add(new_media)
    db.flush()

    # 2️⃣ Create Product record
    product = Product(category_id=ProductCategory_id, reference_id=new_media.id)
    db.add(product)
    db.flush()

    # 3️⃣ Create Quantity record
    qty = Quantity(product_id=product.id, on_hand=0, reserved=0, ordered=0, location=0)
    db.add(qty)
    db.commit()

    return jsonify({
        "message": "Media item created successfully",
        "media": new_media.to_dict(include_relationships=True),
        "product_id": product.id,
        "quantity_id": qty.id
    }), 201


# --- PATCH (partial update) ---
@media_bp.route("/media/<int:id>", methods=["PATCH"])
def update_media(id):
    db = g.db
    item = db.get(Media, id)
    if not item:
        return jsonify({"error": "Media item not found"}), 404

    try:
        data = media_schema.load(request.get_json(), partial=True)
    except ValidationError as err:
        return jsonify({"errors": err.messages}), 400

    for key, value in data.items():
        setattr(item, key, value)

    db.commit()
    return jsonify(media_schema.dump(item)), 200


# --- PUT (full replacement) ---
@media_bp.route("/media/<int:id>", methods=["PUT"])
def replace_media(id):
    db = g.db
    item = db.get(Media, id)
    if not item:
        return jsonify({"error": "Media item not found"}), 404

    try:
        data = media_schema.load(request.get_json())
    except ValidationError as err:
        return jsonify({"errors": err.messages}), 400

    for key, value in data.items():
        setattr(item, key, value)

    db.commit()
    return jsonify(media_schema.dump(item)), 200


# --- DELETE ---
@media_bp.route("/media/<int:id>", methods=["DELETE"])
def delete_media(id):
    db = g.db
    item = db.get(Media, id)
    if not item:
        return jsonify({"error": "Media item not found"}), 404

    # Cascade delete linked product + quantity
    if item.product:
        db.delete(item.product)
    # Also delete linked child product if exists
    if item.child_product:
        db.delete(item.child_product)
    db.delete(item)
    db.commit()
    return jsonify({"message": "Media item deleted successfully."}), 200


# =====================================================
# 🔎 Search Media
# =====================================================
@media_bp.route("/media/search", methods=["GET"])
def search_media():
    db = g.db

    # --- Query parameters ---
    part_number = request.args.get("part_number")
    description = request.args.get("description")
    supplier_name = request.args.get("supplier")
    length = request.args.get("length", type=float)
    width = request.args.get("width", type=float)
    unit_of_measure = request.args.get("unit_of_measure")
    category = request.args.get("category")
    location = request.args.get("location", type=int)

    # Pagination
    page = request.args.get("page", default=1, type=int)
    limit = request.args.get("limit", default=25, type=int)
    offset = (page - 1) * limit

    # --- Base Query ---
    query = (
        select(
            Media.id,
            Media.part_number,
            Media.description,
            Media.length,
            Media.width,
            Media.unit_of_measure,
            Product.id.label("product_id"),
            ChildProduct.id.label("child_product_id"),
            ChildProduct.parent_product_id.label("parent_product_id"),
            Supplier.name.label("supplier_name"),
            MediaCategory.name.label("media_category"),

            Quantity.on_hand,
            Quantity.reserved,
            Quantity.ordered,
            Quantity.location,
            Quantity.available,
            Quantity.backordered,
        )
        .join(Supplier, Media.supplier_id == Supplier.id)
        .join(MediaCategory, Media.category_id == MediaCategory.id)
        .outerjoin(Product, and_(Product.category_id == ProductCategory_id, Product.reference_id == Media.id))
        .outerjoin(ChildProduct, and_(ChildProduct.category_id == ProductCategory_id, ChildProduct.reference_id == Media.id))
        .outerjoin(Quantity, or_(Quantity.product_id == Product.id, Quantity.product_id == ChildProduct.parent_product_id))
        .distinct(Media.id)
    )

    # --- Dynamic Filters ---
    filters = []

    if part_number:
        filters.append(Media.part_number.ilike(f"%{part_number}%"))
    if description:
        filters.append(Media.description.ilike(f"%{description}%"))
    if supplier_name:
        filters.append(Supplier.name.ilike(f"%{supplier_name}%"))
    if length is not None:
        filters.append(Media.length == length)
    if width is not None:
        filters.append(Media.width == width)
    if unit_of_measure:
        filters.append(Media.unit_of_measure.ilike(f"%{unit_of_measure}%"))
    if category:
        filters.append(MediaCategory.name.ilike(f"%{category}%"))
    if location is not None:
        filters.append(Quantity.location == location)

    if filters:
        query = query.where(and_(*filters))

    # --- Total Count ---
    total = len(db.execute(query).mappings().all())

    # --- Pagination ---
    query = query.limit(limit).offset(offset)

    # --- Execute ---
    results = db.execute(query).mappings().all()
    results = [dict(row) for row in results]

    return jsonify({
        "page": page,
        "limit": limit,
        "count": len(results),
        "total": total,
        "results": results
    }), 200

File: frontend/afc-frontend/src/api/media.ts
import { apiRequest } from "./apiClient";

/* ============================================================
   TYPES — match /media/search
============================================================ */

export interface MediaPayload {
  id: number;
  product_id: number;
  child_product_id?: number | null;
  parent_product_id?: number | null;
  part_number: string;
  description?: string | null;
  length: number;
  width: number;
  unit_of_measure: string;

  media_category: string;
  supplier_name: string;

  on_hand: number;
  reserved: number;
  ordered: number;
  available: number;
  backordered: number;
}

export type CreateMediaPayload = {
  part_number: string;
  supplier_id: number;
  category_id: number;
  description?: string | null;
  length?: number;
  width?: number;
  unit_of_measure?: string;
};

export interface MediaResponse {
  page: number;
  limit: number;
  count: number;
  total: number;
  results: MediaPayload[];
}

export interface MediaCategory {
  id: number;
  name: string;
}

/* ============================================================
   SEARCH PARAMS
============================================================ */

export interface MediaSearchParams {
  part_number?: string;
  description?: string;
  supplier?: string;
  category?: string;
  length?: number;
  width?: number;
  unit_of_measure?: string;
  location?: number;
}

/* ============================================================
   API FUNCTIONS
============================================================ */

/**
 * Server-side filtered + paginated media search
 */
export function fetchMedia(
  page = 1,
  limit = 10,
  filters: MediaSearchParams = {}
): Promise<MediaResponse> {
  const params = new URLSearchParams();

  params.set("page", String(page));
  params.set("limit", String(limit));

  if (filters.part_number) params.set("part_number", filters.part_number);
  if (filters.description) params.set("description", filters.description);
  if (filters.supplier) params.set("supplier", filters.supplier);
  if (filters.category !== undefined) params.set("category", String(filters.category));
  if (filters.length !== undefined) params.set("length", String(filters.length));
  if (filters.width !== undefined) params.set("width", String(filters.width));
  if (filters.unit_of_measure) params.set("unit_of_measure", filters.unit_of_measure);
  if (filters.location !== undefined) params.set("location", String(filters.location));

  return apiRequest(`/media/search?${params.toString()}`, {
    method: "GET",
  });
}

export function fetchMediaCategories(): Promise<MediaCategory[]> {
  return apiRequest("/media_categories");
}

/* ============================================================
   CRUD
============================================================ */

export function fetchMediaById(id: string | number) {
  return apiRequest(`/media/${id}`);
}

export function createMedia(data: CreateMediaPayload) {
  return apiRequest("/media", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function updateMedia(
  id: string | number,
  data: Partial<MediaPayload>
) {
  return apiRequest(`/media/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export function patchMedia(
  id: string | number,
  data: {
    supplier_id?: number;
    category_id?: number;
    length?: number;
    width?: number;
    unit_of_measure?: string;
    description?: string | null;
    part_number?: string;
  }
) {
  return apiRequest(`/media/${id}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

export function deleteMedia(id: string | number) {
  return apiRequest(`/media/${id}`, {
    method: "DELETE",
  });
}
